File: examples/typescript/04-control-flow/main.ts
/**
 * Control Flow in TypeScript
 * ==========================
 * TypeScript has the same control flow as JavaScript,
 * plus type narrowing in conditional blocks.
 */

// ── if / else ────────────────────────────────────────

function classify(n: number): string {
    if (n < 0) return "negative";
    if (n === 0) return "zero";
    return "positive";
}

console.log("--- if/else ---");
for (const n of [-3, 0, 7]) {
    console.log(`${n} → ${classify(n)}`);
}

// ── switch ───────────────────────────────────────────

console.log("\n--- switch ---");
function grade(score: number): string {
    const tier = Math.floor(score / 10);
    switch (tier) {
        case 10:
        case 9:  return "A";
        case 8:  return "B";
        case 7:  return "C";
        case 6:  return "D";
        default: return "F";
    }
}

console.log(`grade(95) = ${grade(95)}`);
console.log(`grade(82) = ${grade(82)}`);
console.log(`grade(45) = ${grade(45)}`);

// ── for loop ─────────────────────────────────────────

console.log("\n--- for loop ---");
let total = 0;
for (let i = 1; i <= 5; i++) {
    total += i;
}
console.log(`sum 1..5 = ${total}`);

// ── while loop ───────────────────────────────────────

console.log("\n--- while loop ---");
let x = 1;
while (x < 20) {
    x *= 2;
}
console.log(`first power of two >= 20: ${x}`);

// ── do-while loop ────────────────────────────────────

console.log("\n--- do-while ---");
let j = 0;
do {
    console.log(`j = ${j}`);
    j++;
} while (j < 3);

// ── for-of (iterables) ───────────────────────────────

console.log("\n--- for-of ---");
const colors = ["red", "green", "blue"];
for (const color of colors) {
    console.log(`color: ${color}`);
}

// ── break / continue ─────────────────────────────────

console.log("\n--- break / continue ---");
for (let i = 1; i < 10; i++) {
    if (i % 2 === 1) continue;
    if (i > 6) break;
    console.log(`even <= 6: ${i}`);
}

// ── Type guards (TypeScript-specific) ────────────────

console.log("\n--- type guards ---");

function describe(value: string | number | string[]): string {
    if (typeof value === "string") {
        return `string(${value.length})`;
    } else if (typeof value === "number") {
        return `number(${value})`;
    } else if (Array.isArray(value)) {
        return `array[${value.length}]`;
    }
    return "unknown";
}

console.log(describe("hello"));
console.log(describe(42));
console.log(describe(["a", "b", "c"]));

// ── Summary ──────────────────────────────────────────

console.log("\n--- Summary ---");
console.log("TypeScript control flow:");
console.log("  - if/else for conditionals");
console.log("  - switch for multi-way branching");
console.log("  - for, while, do-while for loops");
console.log("  - break / continue for loop control");
console.log("  - for-of for iterating iterables");
console.log("  - Type guards narrow types in blocks");

File: examples/python/04-control-flow/main.py
"""
Control Flow in Python
======================
Demonstrates if/elif/else, for/while, break/continue, comprehensions,
and match-case (Python 3.10+).
"""

def classify(n: int) -> str:
    if n < 0:
        return "negative"
    elif n == 0:
        return "zero"
    else:
        return "positive"

print("--- if/elif/else ---")
for n in [-3, 0, 7]:
    print(n, "→", classify(n))


print("\n--- for loop (iterables) ---")
total = 0
for i in range(1, 6):
    total += i
print("sum 1..5 =", total)


print("\n--- while loop ---")
x = 1
while x < 20:
    x *= 2
print("first power of two >= 20:", x)


print("\n--- break / continue ---")
for i in range(1, 10):
    if i % 2 == 1:
        continue
    if i > 6:
        break
    print("even <= 6:", i)


print("\n--- comprehensions (expression-style loops) ---")
squares = [i * i for i in range(1, 11)]
evens = [i for i in range(1, 21) if i % 2 == 0]
print("squares:", squares)
print("evens:", evens)


print("\n--- match/case (3.10+) ---")
def describe(value):
    match value:
        case 0:
            return "zero"
        case int() as n if n < 0:
            return f"negative int {n}"
        case int() as n:
            return f"positive int {n}"
        case str() as s:
            return f"string({len(s)})"
        case [a, b]:
            return f"2-list: {a}, {b}"
        case _:
            return "something else"

for v in [0, -5, 7, "hello", [1, 2], {"a": 1}]:
    print(v, "→", describe(v))


# ── Summary ───────────────────────────────────

print("\n--- Summary ---")
print("Python control flow:")
print("  - if/elif/else for conditionals")
print("  - for loop over iterables (lists, strings, ranges)")
print("  - while loop")
print("  - break / continue for loop control")
print("  - list comprehensions for concise loops")
print("  - match/case for pattern matching (3.10+)")
